### ISSUE TITLE
Update Django within current Python 2 constrains

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2254]
With reference to "move to python 3" #1877 it is proposed that we begin our Django update 'train' from the current just prior LTS version (1.8.16) to the next proper LTS version of 1.8.19 and then in turn to the last Python 2 supported version of 1.11.29.

This issue it intended to cover this proposed update prior to our Python 3 move as newer versions of Django than 1.11.29 will have to wait until our base code is transitioned to Python 3.

In summary:
1.8.16 -> 1.8.19 so we are on LTS.
Then:
1.8.19 -> 1.11.29 to reach the latest available for Python2 via an LTS to LTS update.

A future issue/pr can then take us to Django 2 or 3 once we have #1877 sorted.

Due to the potentially deep and broad changes that may be required in this update it is further proposed that all work take place in parallel to our current master branch by way of a new testing branch. Once all our tests pass and all looks to be good the work can then be rebased on master and merged accordingly. This should help with avoiding breakage in what is currently our Release Candidate branch of master. In the future, given the scale of our technical debt, it may be that we should move to using the proposed testing branch to do all of the larger pull requests to come and release our testing rpms of the future from this testing branch. Leaving the possibility to have master as the Stable release branch. This arrangement is currently only a proposal however.

**Comment by phillxnet** [comment]
I have begun looking into this issue. It remains to be seen if this is an appropriate issue for the allocated milestone.
Fingers crossed.

**Comment by phillxnet** [comment]
It would appear that in moving to 1.11.29 we are also loosing compatibility with our current django-rest-framework version 3.1.1.

> jangorestframework-3.1.1-py2.7.egg/rest_framework/response.py", line 8, in <module>', '    from django.core.handlers.wsgi import STATUS_CODE_TEXT', 'ImportError: cannot import name STATUS_CODE_TEXT', '']

https://docs.djangoproject.com/en/1.11/releases/1.9/#miscellaneous

> django.http.responses.REASON_PHRASES and django.core.handlers.wsgi.STATUS_CODE_TEXT have been removed. Use Python’s stdlib instead: http.client.responses for Python 3 and httplib.responses for Python 2.

https://docs.python.org/2/library/httplib.html#httplib.responses

**Comment by phillxnet** [comment]
## django-rest-framework
https://www.django-rest-framework.org/#requirements
states Python 3 and Django 2.2 + only for current versions!

Last version to support Python 2 was 3.9.3:
https://www.django-rest-framework.org/community/release-notes/#393
Note that 3.9.1 added the following additional requirement:

> Upgrade Bootstrap to 3.4.0 to resolve XSS issue.

**Comment by phillxnet** [comment]
While updating urls.py to comply with the newer Django format it looks like the the embedded urls within our oath plugin:

django_oauth_toolkit-0.9.0

are similarly old and now removed format. I.e. using the now removed in Django 1.10 patterns import:

> File "/opt/rockstor-dev/eggs/django_oauth_toolkit-0.9.0-py2.7.egg/oauth2_provider/urls.py", line 2, in <module>', '    from django.conf.urls import patterns, url', 'ImportError: cannot import name patterns', '']

**Comment by phillxnet** [comment]
## django_oauth_toolkit
Another Python 3 only in current form:
https://github.com/jazzband/django-oauth-toolkit#requirements
- Python 3.5+
- Django 2.1+
- oauthlib 3.1+

https://github.com/jazzband/django-oauth-toolkit/blob/master/CHANGELOG.md#110-2018-04-13
### 1.1.0 [2018-04-13]
- Compatibility: Django 1.11 is the new minimum required version. Django 1.10 is no longer supported.
- Compatibility: This will be the last release to support Django 1.11 and Python 2.7.

**Comment by phillxnet** [comment]
### requests update required as django_oauth_toolkit-1.1.2 requires at least 2.13

> Getting distribution for 'django-oauth-toolkit==1.1.2'.
> Got django-oauth-toolkit 1.1.2.
> Version and requirements information containing requests:
>   Requirement of rockstor: requests==1.1.0
>   Requirement of django-oauth-toolkit==1.1.2: requests>=2.13.0
> While:
>   Updating django.
> Error: There is a version conflict.
> We already have: requests 1.1.0
> but django-oauth-toolkit 1.1.2 requires 'requests>=2.13.0'.

## Requests
https://requests.readthedocs.io/en/master/
Looks to have ongoing Python 2.7 support!
"Requests officially supports Python 2.7 & 3.5+, ..."

**Comment by phillxnet** [comment]
## chardet update

> Installing django.
> Getting distribution for 'requests==2.25.1'.
> Got requests 2.25.1.
> Version and requirements information containing chardet:
>   Requirement of rockstor: chardet==2.3.0
>   Requirement of requests==2.25.1: chardet<5,>=3.0.2
> While:
>   Installing django.
> Error: There is a version conflict.
> We already have: chardet 2.3.0
> but requests 2.25.1 requires 'chardet<5,>=3.0.2'.

## Chardet
https://pypi.org/project/chardet/
"Requires Python 2.7 or 3.5+."
However we have:
https://github.com/chardet/chardet/commit/4291399777cd6f011bfdab5da678ecbea8f10855
"Remove reference to Python 2"  16 Dec 2020
https://github.com/chardet/chardet

**Comment by phillxnet** [comment]
## django_ztask
Upstream last updated on 6 Jun 2012 !!

> Feb 02 13:59:01 rleap15-2 initrock[1450]: hon2.7/importlib/__init__.py", line 37, in import_module', '    __import__(name)', '  File "/opt/rockstor-dev/eggs/django_ztask-0.1.5-py2.7.egg/django_ztask/migrations/0001_initial.py", line 3, in <module>', '    from south.db import db', 'ImportError: No module named south.db', '']

https://github.com/dmgctrl/django-ztask/issues/13
No current maintainer.

_Looking for maintained (newer Django/Python 3 compatible) drop-in replacements. At least for now._

**Comment by phillxnet** [comment]
## Async / background Django replacement
So it seem there is no currently maintained drop-in replacement for django_ztask.

Django 3.0 lays the ground work for native Django support but we have a way to go before this can used. Plus we need a solution to see us through this phase of dependency upgrades. I.e. Django 1.11 / 2.0 Python 2.7/3 transitions.
And we have the chicken/egg scenario re Python 3 async. Also note there that our current django_ztask clashes with the namespaces used in Python 3. Though we hardly going to be mixing these two anyway.

Researching replacement to fit our current and future needs:
Relatively recent article on async Django tasks: https://stuartm.com/2020/05/django-async/

Our use of ztask async is predominantly in our config save/restore, Rock-ons management, and longer running btrfs tasks such as scrub and balance:
```
grep -R '@task'
src/rockstor/fs/btrfs.py:@task()
src/rockstor/fs/btrfs.py:    bypasses our @task() decorator and we are then called directly.
src/rockstor/storageadmin/views/rockon_helpers.py:@task()
src/rockstor/storageadmin/views/config_backup.py:@task()
src/rockstor/smart_manager/views/ztask_helpers.py:@task()
```
and:
```
grep -R '\.async'
rockstor/storageadmin/views/pool.py:        start_balance.async(mnt_pt, force=force, convert=convert)
rockstor/storageadmin/views/pool.py:        start_resize_pool.async(cmd)
rockstor/fs/btrfs.py:    without the async function extension (start_resize_pool.async()) which
rockstor/smart_manager/views/rockstor_service.py:                    ztask_helpers.restart_rockstor.async(nco.ipaddr, listener_port)
...
```

**Comment by phillxnet** [comment]
## Version options for considered background task managers

### Celery
Complex and possibly over fill. Also Celery is python, but needs rabbitmq which needs erlang!

### Django-Q
https://github.com/Koed00/django-q/blob/master/CHANGELOG.md
Unclear when/if Python 2.7 is still supported.
https://pypi.org/project/django-q/
https://django-q.readthedocs.io/en/latest/install.html#requirements

> "Since we are now no longer supporting Python 2, we can also not support older versions of Django that do not support Python >= 3.6 For this you can always use older releases, but they are no longer maintained."

**Comment by phillxnet** [comment]
@FroggyFlox On the question of the current 'blocker' dependency of django-ztasks and it's replacement; I'm currently favouring Huey as it looks to be both Python 2.7 and 3 and have older Django support, if we use an old/new enough version. This paves the way for it meeting our immediate and future needs in this transition phase (hopefully). More reading to be done as yet though.

**Comment by phillxnet** [comment]
The preferred Huey option has also attracted the following young but promising django admin addon project:
https://pypi.org/project/django-huey-monitor/
https://github.com/boxine/django-huey-monitor
It is not proposed that we integrate this as well, at least until we have our Django and Python updates fully in place and maintained on a regular basis.

**Comment by phillxnet** [comment]
This issue, and it's milestone status, is currently dependant on the in-progress and linked:
"Replace orphaned django-ztask with Huey" #2276

**Comment by phillxnet** [comment]
Now we have the django-ztask to Huey move resolved via pr:
"Replace orphaned django-ztask with Huey (#2276)" #2283
and released in the testing channel from version 4.0.6 I will re-base my progress there to date and continue with this issue shortly ; hopefully.

**Comment by phillxnet** [comment]
Removed GitHub branch associated with this ongoing work and rebased on current master - tag 4.0.7-0.
Will re-push and continue from there so that we can inherrit the required, and referenced, ztasksd-to-Huey move and all other ongoing changes since this issues associated work was started.

**Comment by phillxnet** [comment]
OK so a way to go yet and no working login but we at least now have some tests passing:
```
Ran 207 tests in 5.613s

FAILED (failures=46, errors=28)
```

**Comment by phillxnet** [comment]
Progress of sorts !!
```
Ran 207 tests in 5.837s

FAILED (failures=2, errors=43)
```

**Comment by phillxnet** [comment]
Definite progress:
```
Ran 207 tests in 15.257s

FAILED (failures=4, errors=10)
```

**Comment by phillxnet** [comment]
The current state of this branch, i.e. a few tests still failing and a tricky db update issue, along with the many core package updates associated with the Django update and the already hundreds of changes to date, leads me to removing it's optimistic inclusion in our current ISO v4 release milestone. I will likely use this pr instead to establish a new testing branch and release sequence so that we can then begin again in our move towards the next stable release post 4.0.x.

**Comment by phillxnet** [comment]
> I will likely use this pr instead to establish a new testing branch and release sequence so that we can then begin again in our move towards the next stable release post 5.x.x.

@Hooverdan96 's recent explorations in this area have encouraged me to move forward on at least getting this work into our newly establish testing channel. The above failing 14 unit tests can then be more easily addressed via a team effort and the sited reboot db migration issue, post initially successful migration, can in turn be approached within their own testing channel specific issues.

I propose that we have say a [t] indicator for all issues that pertain to the testing channel going forward maybe.
i.e.
[t] Unit tests result in FAILED (failures=4, errors=10)
via
```
cd <root dir of rockstor ie /opt/rockstor-dev>
./bin/test -v 2 -p test_*
```

**Comment by phillxnet** [comment]
Consequently my next task here is to re-base on top of our existing stable (master) branch and push this to testing to encourage more collaboration in the greater issue of moving up to Python 3.

**Comment by phillxnet** [comment]
Closing as
fixed by #2365
and some more minor follow up fixes in that area. We are now, in testing branch, using the last Python 2.7 compatible Django available.
